apache/spark
# [WIP][SPARK-42240][INFRA][CONNECT][TESTS] Move `ClientE2ETestSuite` into a separate module and add new GA task to test shaded jvm client with maven

### What changes were proposed in this pull request?
The main change of this pr as follows:
- Move `ClientE2ETestSuite` into a new separate module, the new module named `connect-jvm-e2e-tests`, and add a new profile named `connect-client-tests` to activate this module. The new profile is added because the current test way of this module is cumbersome and it is not easy to build and test by default. 

we need the follow steps for the new module test with maven

```
build/mvn clean -Pscala-2.12
build/mvn -DskipTests install -pl repl -am -Pscala-2.12
build/mvn -DskipTests install -pl connector/connect/common -Pscala-2.12
build/mvn -DskipTests install -pl connector/connect/server -Pscala-2.12
build/mvn -DskipTests install -pl connector/connect/client/jvm -Pscala-2.12
build/mvn clean test -pl connector/connect/client/jvm-e2e-tests -Pscala-2.12 -Pconnect-client-tests
```

or 

```
build/mvn clean -Pscala-2.12
build/mvn -DskipTests clean install -Pscala-2.12
build/mvn clean test -pl connector/connect/client/jvm-e2e-tests -Pscala-2.12 -Pconnect-client-tests
```

It seems simpler to use sbt for test, but in the current dependency management way, using sbt to fetch the no-shaded connect-client jar cannot achieve the purpose of testing both the shaded client jar and the shaded server jar in the E2E tests.


```
build/sbt clean package -Pscala-2.12
build/sbt "connect-client-jvm-e2e-tests/test" -Pscala-2.12 -Pconnect-client-tests
```


- Add a new GA task named `connect-jvm-e2e-tests`, this task use maven to test shaded `connect-client-jvm`. Under the current dependency management way, if we can find out how to fetch the assembly jar by sbt, we should change it to a more efficient sbt test mode.

- Remove `repl` test dependency from `connector/connect/client/jvm/pom.xml` due to `ClientE2ETestSuite` move into separate module.


### Why are the changes needed?
We should do E2E test with shaded client jar and shaded server jar. Before this pr, `ClientE2ETestSuite` is in `connect-client-jvm` module and the test will using shaded server jar and no-shaded client jar. After this pr, the test purpose can be achieved by using maven, but the test way using sbt needs to be further explored.



### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
- Pass GitHub Actions
- New GA Task passed
- Manual check:

Scala 2.13 using maven 

```
gh pr checkout 39807
dev/change-scala-version.sh 2.13
build/mvn clean -Pscala-2.13
build/mvn -DskipTests install -pl repl -am -Pscala-2.13
build/mvn -DskipTests install -pl connector/connect/common -Pscala-2.13
build/mvn -DskipTests install -pl connector/connect/server -Pscala-2.13
build/mvn -DskipTests install -pl connector/connect/client/jvm -Pscala-2.13
build/mvn  clean test -pl connector/connect/client/jvm-e2e-tests -Pscala-2.13 -Pconnect-client-tests
```

Scala 2.12 using sbt 


```
gh pr checkout 39807
build/sbt clean package -Pscala-2.12
build/sbt "connect-client-jvm-e2e-tests/test" -Pscala-2.12 -Pconnect-client-tests
```

Scala 2.13 using sbt 

```
gh pr checkout 39807
dev/change-scala-version.sh 2.13
build/sbt clean package -Pscala-2.13
build/sbt "connect-client-jvm-e2e-tests/test" -Pscala-2.13 -Pconnect-client-tests
```
